flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -513,19 +514,20 @@ private void increaseBuffersInBacklog(BufferConsumer buffer) {
         }
     }
 
-    /**
-     * Gets the number of non-event buffers in this subpartition.
-     *
-     * <p><strong>Beware:</strong> This method should only be used in tests in non-concurrent access
-     * scenarios since it does not make any concurrency guarantees.
-     */
-    @SuppressWarnings("FieldAccessNotGuarded")
-    @VisibleForTesting
+    /** Gets the number of non-event buffers in this subpartition. */
     public int getBuffersInBacklog() {
-        if (flushRequested || isFinished) {
-            return buffersInBacklog;
-        } else {
-            return Math.max(buffersInBacklog - 1, 0);
+        synchronized (buffers) {
+            if (isBlocked || buffers.isEmpty()) {
```

Comment:
After the channel is blocked, we do not need to allocated any credit at the downstream, because no buffer will be sent. As mentioned in your comment, if the initialCredit is 0, it should return 0. If the initialCredit is not 0, we always have exclusive credits, we still do not need allocate any floating credit. Maybe returning the actual value instead of 0 is a small improvement.